src/swaps/jupiter/JupiterSwapProvider.ts

```diff
@@ -0,0 +1,243 @@
+  }
+
+  public async newSwap(quoteInput: SwapRequest<JupiterSwapHistoryItem>) {
+    const connection = new Connection(
+      'https://solana--mainnet.datahub.figment.io/apikey/d7d9844ccf72ad4fef9bc5caaa957a50',
+      'confirmed'
+    );
+    const { network, walletId, quote } = quoteInput;
+    const client = this.getClient(network, walletId, quote.from, quote.fromAccountId);
+    const [{ address }] = await client.wallet.getAddresses();
+
+    const transactions = await this._getTransactions(quote.info, address);
+
+    let swapTx;
+    const txTypes = Object.keys(transactions);
+
+    /* Full swap flow might be constructed from 3 different transactions:
+      1. Setup transaction  - used to handle creating ATA or Serum open order accounts
+      2. Swap transaction  - performing the actual swaps
+      3. Clean up transaction - unwrap SOL if a SOL swap.
+      You must send each transaction sequentially in the order:
+      setupTransaction -> swapTransaction -> cleanupTransaction and wait for each to be 'confirmed' before sending the next one.
+    */
+
+    for (let i = 0; i < txTypes.length; i++) {
+      const tx = Transaction.from(Buffer.from(transactions[txTypes[i]], 'base64'));
+
+      if (txTypes[i] === 'swapTransaction') {
+        // @ts-ignore TODO: CAL should allow `any` for data
+        swapTx = await client.wallet.sendTransaction({ transaction: tx, value: new BN(quote.fromAmount) });
+        await connection.confirmTransaction(swapTx.hash);
+      } else {
+        // @ts-ignore TODO: CAL should allow `any` for data
+        const transaction = await client.wallet.sendTransaction({ transaction: tx, value: new BN(0) });
+        await connection.confirmTransaction(transaction.hash);
+      }
+    }
+
+    const updates = {
+      status: 'WAITING_FOR_SWAP_CONFIRMATIONS',
+      swapTx,
+      swapTxHash: swapTx?.hash,
+    };
+
+    return {
+      id: uuidv4(),
+      fee: quote.fee,
+      slippage: 50,
+      ...updates,
+    };
+  }
+
+  public async estimateFees({
+    txType,
+    feePrices,
+    asset,
+  }: EstimateFeeRequest<string, JupiterSwapHistoryItem>): Promise<EstimateFeeResponse | null> {
+    if (txType != this.fromTxType) {
+      throw new Error(`Invalid tx type ${txType}`);
+    }
+
+    const nativeAsset = chains[cryptoassets[asset].chain].nativeAsset;
+
+    const gasLimit = 1000000000;
+    const fees: EstimateFeeResponse = {};
+
+    for (const feePrice of feePrices) {
+      const fee = new BN(gasLimit).times(feePrice);
+      fees[feePrice] = new BN(unitToCurrency(cryptoassets[nativeAsset], fee).toFixed());
+    }
+
+    return fees;
+  }
+
+  // Jupiter on their site before performing swap are recommending:
+  // "We recommend having at least 0.05 SOL for any transaction"
+  // https://jup.ag/swap/SOL-USDT
+
+  // Here we are returning `33333333` Lamports because in wallet we multiply by 1.5 which
+  // output exactly 0.05 SOL
+  getExtraAmountToExtractFromBalance() {
+    return 33333333;
```

Comment:
this doesn't look good... what's the idea here?

Reply:
described in the comment above the function

> Also any jupiter testnet?

no :(

Reply:
<img width="450" alt="image" src="https://user-images.githubusercontent.com/32274987/181742632-8d5223ef-6b34-4ff2-a853-bb839cf63522.png">
 They recommend to have this amount for possible fees available, comment above the function

Reply:
33333333 * 1.5 = 49999999,5 😆 

where is the multiplication of 1.5? can we skip it and set 0.05 here instead?

Reply:
Yes we add to this fee which is static and it is - 00000000.5 SOL,

Here is the multiplyer: 

https://github.com/liquality/wallet/blob/2ca93e41dfc52efe5f55659e1b14956849f17b6e/src/views/Swap/Swap.vue#L884

Reply:
Maybe i can pass some boolean to the wallet and skip multiplication for Solana chain not sure if this is good tho

Reply:
i think its better to have some leftover then

this is only when sending max, right?

Reply:
This left over is kept every time, let's say you have 2 SOL, Max available will be 1.95

Reply:
If you want to send just 1 SOL, no problem you send 1, just reducing the max amount and keep some for extra fee if needed

Reply:
I think there are 2 reasons for this:

1. Jupiter is not aware how many txs will be broadcasted 1-3
2. Tx for creation of token accounts, which are part of this jupiter transactions